include/dlaf/eigensolver/tridiag_solver/permutations.h

```diff
@@ -0,0 +1,108 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+#pragma once
+
+#include <vector>
+
+#include "dlaf/blas/tile.h"
+#include "dlaf/matrix/matrix.h"
+#include "dlaf/types.h"
+
+namespace dlaf::eigensolver::internal {
+
+// Applies the permutaton index `perm_arr` to a portion of the columns/rows(depends on coord) [1] of an
+// input submatrix [2] and saves the result into a subregion [3] of an output submatrix [4].
+//
+// Example column permutations with `perm_arr = [8, 2, 5]`:
+//
+//          2     5     8      out_begin
+//     ┌────────────────────┐    ┌─────┬─────────────┐
+//     │  in_offset      in │    │     │         out │
+//     │ │                  │    │     └─►┌─┬─┬─┐    │
+//     │◄┘ ┌─┐   ┌─┐   ┌─┐  │    │        │c│a│b│    │
+//     │   │a│   │b│   │c│  │    │    ┌──►│ │ │ │    │
+//     │   │ │ ┌►│ │   │ │◄─┼────┼──┐ │   └─┴─┴─┘    │
+//     │   └─┘ │ └─┘   └─┘  │    │  │ │      ▲       │
+//     │       │            │    │sz.rows()  │       │
+//     │      sz.rows()     │    │           └─      │
+//     │                    │    │          sz.cols()│
+//     └────────────────────┘    └───────────────────┘
+//
+// Example row permutations with `perm_arr = [3, 1]`:
+//
+//             ┌─── in_offset
+//             │                   out_begin
+//     ┌───────▼────────────┐    ┌──┬────────────────┐
+//     │                 in │    │  │            out │
+//     │       ┌─────┐      │    │  └►┌─────┐        │
+//   1 │       │  a  │      │    │    │  b  │        │
+//     │       └─────┘      │    │    ├─────┤ ◄─┐    │
+//     │                    │    │    │  a  │   │    │
+//     │       ┌─────┐      │    │    └──▲──┘   │    │
+//   3 │       │  b  │      │    │       │      │    │
+//     │       └──▲──┘      │    │ sz.cols()    │    │
+//     │          │         │    │          sz.rows()│
+//     └──────────┼─────────┘    └───────────────────┘
+//                │
+//             sz.cols()
+//
+//
+// [1]: The portion of each input column or row is defined by the interval [in_offset, in_offset +
+//      sz.col()) or the interval [in_offset, in_offset + sz.row()) respectively.
+// [2]: The input submatrix is defined by `begin_tiles`, `ld_tiles`, `distr` and `in_tiles`
+// [3]: The subregion is defined by `begin` and `sz`
+// [4]: The output submatrix is defined by `begin_tiles`, `ld_tiles`, `distr` and `out_tiles`
+//
+// clang-format off
+template<class T, Coord coord>
+void applyPermutations(
+    GlobalElementIndex out_begin,
+    GlobalElementSize sz,
+    SizeType in_offset,
+    const matrix::Distribution& distr,
+    const SizeType* perm_arr,
+    const std::vector<matrix::Tile<T, Device::CPU>>& in_tiles,
+    std::vector<matrix::Tile<T, Device::CPU>>& out_tiles
+    ) {
+  // clang-format on
+
+  // Iterate over rows if `coord == Coord::Row` otherwise iterate over columns
+  SizeType dim1 = sz.cols();
+  SizeType dim2 = sz.rows();
+  if constexpr (coord == Coord::Row)
+    std::swap(dim1, dim2);
```

Comment:
Does swapping the dimensions like this have any effect on the access pattern into the tile? Not a blocker and can be left as future optimization, but just curious now.